http/peer-id-auth.md

```diff
@@ -0,0 +1,194 @@
+| challenge-client | The random base64 encoded value the server generates to challenge the client to prove its identity                                                                                                                                       |
+| sig              | A base64 encoded signature.                                                                                                                                                                                                              |
+| peer-id          | The peer id of the node that set this parameter. Encoding defined by the [Peer ID spec].                                                                                                                                                 |
+| public-key       | A base64 encoded value of peer's public key. The key itself is encoded per the [Peer ID spec].                                                                                                                                           |
+| opaque           | An base64 encoded opaque to the client blob generated by the server. If a client receives this it must return it. A server may use this to authenticate statelessly. For example, it could store the challenge-client and a expiry time. |
+
+Params are encoded per [RFC 9110 auth-param's ABNF](https://datatracker.ietf.org/doc/html/rfc9110#name-collected-abnf). Generally it'll be something like: `hostname="example.com", challenge-server=base64EncodedVal`
+
+## Signing
+
+Signatures sign some set of parameters. The parameters are sorted
+alphabetically, prepended with a varint length prefix, and concatenated together
+to form the data to be signed. The signing algorithm is defined by the key type
+used. Refer to the [Peer ID
+spec] for
+specifics on the signing algorithm. The set of parameters is prefixed with the auth scheme "libp2p-PeerID"
+
+As an example, if we wanted to sign the parameters `hostname="example.com",
+challenge-client=base64String` we would first structure the parameters as a byte
+slice containing:
+```
+libp2p-PeerID<varintprefix>challenge-client=<base64String><varintprefix>hostname="example.com"
+```
+
+Then sign the resulting byte slice. See the test vectors below for a
+examples.
+
+
+## Base64 Encoding
+
+The base64 encoding follows Base 64 Encoding with URL and Filename Safe Alphabet
+from [RFC 4648](https://datatracker.ietf.org/doc/html/rfc4648#section-5). The
+reason this is not a multibase is to aid clients or servers who can not or
+prefer not to import a multibase dependency.
+
+## Mutual Client and Server Peer ID
+
+The following protocol allows both the client and server to authenticate each
+other's peer id by having them each sign a challenge issued by the other. The
+protocol operates as follows:
+
+1. The client makes a POST request to the authentication URI.
```

Comment:
Does this need to be the authentication URI? Can this work for any resource with the server responding with a 401 and a challenge and client doing a retry of the request with the auth headers?

Reply:
It could, but that's a bit trickier. I think the simpler flow is:

1. client does mutual auth with the authentication URI.
2. client makes a request to the authenticated endpoint with the bearer token.

That said, I don't want to forbid it. I'll try to change the language to allow it while also recommending this simpler strategy.